rfcs/0085-nixos-release-stablization.md

```diff
@@ -0,0 +1,128 @@
+---
+feature: nixos-release-stablization
+start-date: 2021-01-17
+author: Jonathan Ringer (@jonringer)
+co-authors:
+shepherd-team: @ryantm, @garbas, @Mic92
+shepherd-leader: @ryantm
+related-issues: [NixOS release schedule](https://github.com/NixOS/rfcs/pull/80)
+---
+
+# Summary
+[summary]: #summary
+
+To bring more certainty to the release cycle, add short periods where
+breaking changes are partially restricted. A list of Release Critical
+Packages is defined. Also, move most release stablization work from
+the `release` branch to the `master` branch to reduce backports.
+
+# Motivation
+[motivation]: #motivation
+
+We want to release on time without herculian efforts. Limited
+restrictions to merging breaking changes should avoid the common
+practice of forcing many, largely-untested, commits into master before
+branch-off.
+
+This happened for the 20.09 release in which ~640 staging
+commits were pushed through in the days leading up to the branch off
+([#95492](https://github.com/NixOS/nixpkgs/pull/95492),
+[#96280](https://github.com/NixOS/nixpkgs/pull/96280),
+[#96437](https://github.com/NixOS/nixpkgs/pull/96437),
+[#97146](https://github.com/NixOS/nixpkgs/pull/97146)). The last PR included a
+release blocker fix (introduced in [one of the mentioned
+PRs](https://github.com/NixOS/nixpkgs/pull/96437)) which delayed the
+branch-off from 4 Sept 2020 to 7 Sept 2020, and ZHF had to be delayed
+another day to get a semi-accurate hydra build status. The systemd
+bump was also largely responsible delaying the release until after the
+scheduled release date where basic login features in plasma [weren't
+restored until Oct 5th, a week after the planned release
+date](https://github.com/NixOS/nixpkgs/pull/99629) and a backlog of
+other plasma related issues [weren't resolved until Oct 19th, almost a
+month after the planned release
+date](https://github.com/NixOS/nixpkgs/pull/101078).  This is not to
+discredit any individual, but demonstrate that certain packages have
+many dimensions of compatibility which need a longer time to stabilize
+on unstable before being included in a release.
+
+
+# Definitions
+[definitions]: #definition
+- Breaking change
+  - A change that is likely to break downstream.
+  - For projects following Semantic Versioning, disallow major version bumps.
```

Comment:
Do I read this right that it would still allow packages that have to be on a specific version to be useful (steam, spotify, slack, ...) because a remote server requires it are not affected by this rule even if they are major version upgrade? This scenario would probably qualify as bug fix and thus be fine?

Reply:
I think we already have an unwritten policy for projects like discourse/skype/steam/spotify that allows backports to stable, even for major version to stay compatible with third parties.

Reply:
I am aware of that but now that we start introducing a written policy we might want to mention that packages that only work in combination with a remote service might be excluded.

Reply:
It's the defining the term. The "restricted" updates only apply to staging. So if the major version bump affects <100 (or 500 in stituations, e.g. python packages) rebuilds, then it's still free to be bumped. All PRs which target master will not be affected by any of these changes.

Reply:
Please refer to https://github.com/jonringer/rfcs/blob/85-ZHF-on-master/rfcs/0085-nixos-release-stablization.md#restriction-timeline, where "restrictions" are only placed on the staging and staging-next branches.

The goal is just to "limit potential breakages" and creating more work trying to stabilize master. Until we have a great process for stabilizing staging-next, this is the best compromise I could think of.

However, the current `staging-next` process has many short comings:
- Mostly core contributors are stabilizing it, and unable to dedicate time to fix all breakages
- The goal is for it to be "stable enough", and not necessarily be green-to-green

For ZHF and prevent trying to fix previously working packages, just disallowing updates which are likely to cause regressions seems like the most prudent course of action.

Reply:
The applications in question (steam, spotify, discord, slack, etc. ) will generally target master, and can be backported. This RFC does not affect these use cases.